### PR TITLE
Support entry format for date component

### LINKED ISSUE #1850: support entry format for dateand time in questionnaire.
**Describe the bug**
Developer should define the date format for the questionnaire.
More details are available in  https://github.com/google/android-fhir/issues/1802  as this issue is one part of  issue #1802

**Component**
 SDC library

### PR DESCRIPTION
**Pull request by PallaviGanorkar**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1850

**Description**
1. supported entry format in questionnaire for date component .
2. If entry format not present in questionnaire then device locale date format is used

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Feature

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DatePickerViewHolderFactory.kt [hunk 1]
```diff
@@ -282,6 +284,16 @@ internal fun handleDateFormatAfterTextChange(
 internal const val TAG = "date-picker"
 internal val ZONE_ID_UTC = ZoneId.of("UTC")
 
+internal fun datePattern(questionnaireItem: Questionnaire.QuestionnaireItemComponent): String {
```
[the remark this answers is not in the document]
Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DatePickerViewHolderFactory.kt:287]: Here entryFormat extension is used to read value from QuestionnaireItemComponent if its value is not available or not as expected then this method return device local date pattern which is not part QuestionnaireItemComponent. So i have preferred to write this functionality as separate method.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DatePickerViewHolderFactory.kt [hunk 2]
```diff
@@ -282,6 +284,25 @@ internal fun handleDateFormatAfterTextChange(
...
+        getLocalizedDateTimePattern()
+      } catch (e: DateTimeParseException) {
+        getLocalizedDateTimePattern()
+      } catch (e: Exception) {
+        getLocalizedDateTimePattern()
+      }
+    }
+      ?: getLocalizedDateTimePattern()
+  return str
+}
```
Comment (aditya-07) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DatePickerViewHolderFactory.kt:304]:
```suggestion
internal fun datePattern(questionnaireItem: QuestionnaireItemComponent) =
  if (isValidDateEntryFormat(questionnaireItem.entryFormat)) {
    questionnaireItem.entryFormat
  } else {
    getLocalizedDateTimePattern()
  }

fun isValidDateEntryFormat(entryFormat: String?): Boolean {

  if (entryFormat == null) return false
  return try {
    val text = LocalDate.now().format(DateTimeFormatter.ofPattern(entryFormat))
    LocalDate.parse(text, DateTimeFormatter.ofPattern(entryFormat))
    true
  } catch (e: Exception) {
    false
  }
}
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreLocalDates.kt [hunk 3]
```diff
@@ -31,7 +31,7 @@ import java.util.Date
  * "dd/mm/yyyy").
  */
 internal fun getDateSeparator(localeDatePattern: String): Char =
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreLocalDates.kt:33]: as discussed, the right thing to do here is probably make the return type nullable?

**Review by jingtang10 — CHANGES_REQUESTED** [review]

**Comment by jingtang10** [comment]
I did some tidying up:

- reorganized MoreQuestionnaireItemComponents.kt file and put things in groups
- added unit tests
- merged to HEAD

**Review by aditya-07 — APPROVED** [review]
LGTM

**Review by jingtang10 — APPROVED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DatePickerViewHolderFactory.kt [hunk 4]
```diff
@@ -101,17 +98,19 @@ internal object DatePickerViewHolderFactory :
...
       override fun bind(questionnaireViewItem: QuestionnaireViewItem) {
         clearPreviousState()
         header.bind(questionnaireViewItem)
+
+        val datePattern = questionnaireViewItem.questionnaireItem.dateEntryFormatOrSystemDefault
+        // Special character used in date pattern
+        val datePatternSeparator = getDateSeparator(datePattern)
+        textWatcher = DatePatternTextWatcher(datePatternSeparator)
+        canonicalizedDatePattern = canonicalizeDatePattern(datePattern)
+
```
Comment (FikriMilano) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/factories/DatePickerViewHolderFactory.kt:113]: this changes broke the catalog app's behaviour calculated expression.
the widget cannot update the date based on calculated-expression from the quantity widget below it.

if you move this block of code to init() the widget starts working again, but we cannot use dateEntryFormat in that case.

**Pull request merged**